### ISSUE TITLE
Marvin import is slow

### ISSUE THREAD

**Issue by albireox** [sdss/marvin#442]
**Describe the bug**
Importing marvin (`import marvin`) takes 3.5 seconds

```
CPU times: user 3.01 s, sys: 245 ms, total: 3.26 s
Wall time: 3.54 s
```

This seems way too long. Why is it so slow? Running it without internet takes the same so it's not likely to be related to an API request. Without a PostgreSQL server it goes down to 2 seconds, which is still too much. Also, without a SQL server I get this warning that I usually don't get

```
/Users/albireo/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/lib/python3.7/site-packages/sqlalchemy/ext/declarative/clsregistry.py:120: SAWarning: This declarative base already contains a class with the same class name and module name as marvin.db.models.SampleModelClasses.MangaTarget, and will be replaced in the string-lookup table.
  item.__name__
```

Is this something that users normally see?

**Expected behaviour**
Faster-than-light marvin import.

**Comment by havok2063** [comment]
Can you time the pip-installed user-experience version of Marvin?  We talked about this before and last time you deemed it as ok.  I agree it is slow for us, the developers.

**Comment by albireox** [comment]
On a new virtualenv, with my PortgreSQL server stopped, and installing from pip I still get 1.8s. It's not terrible but a bit annoying and would be good to know why. It's not critical but I'll try some profiling and see if there is something very obvious that we can trim.

**Comment by albireox** [comment]
Ok. I found that the main cause of the slowdown (and I think something new since the last time we checked this) is that the maskbits are being loaded to the datamodel on import. To create the maskbits, the yanny file needs to be read multiple times, which is slow. A one-line change can make marvin read the yanny file only once and brings down the import to 0.9s. Note that this is not the nicest fix but I think I'll create a PR because at least it's a good gain.

A few other things we should consider at some point:
- We import way too many things on init that arguably are not needed at that point. Some may be very complicated to untangle at this point but others may not be that complicated.
- The datamodel should not need to be imported on load, and even without the yanny file reads is still a small part of what makes things slow.
- The other big slowdown (0.3s) is the import of the modelclasses / db stuff even when there isn't a database. `MarvinDB` basically goes through all the motions even if there is no connection and it try-excepts them, but my guess is that before failing they consume a fair amount of CPU time. It's probably possible to realise earlier on that there is no DB and just stop the whole process.

**Comment by havok2063** [comment]
Just now seeing your comments here.  Yeah I think the ModelClasses and DataModel behaviour can be adjusted but we need to be a bit careful about how that gets reorganized.  There is some circularity we introduce by moving things around too much, that we'd need to sort out.  We can probably implement some internal caching of how `MarvinDB` handles itself.  I'll create a new Issue for this specific thing.

**Comment by albireox** [comment]
Sounds good. As a note for future efforts on this, I'd say the objective should be that we can remove `sqlalchemy`, `SQLAlchemy-boolean-search`, and `wtforms-alchemy` from the requirements/external for normal users. In principle there is no reason why a user without a manga DB should need to have sqlalchemy installed to use Marvin or the query. I *think* right now we only use it for `SQLAlchemy-boolean-search` but really that is just a check of the filter and it is not necessary to perform the query (which happens remotely). That said, I don't know that part of the code that well and I may be missing something.

**Comment by havok2063** [comment]
Regarding the timing of the import in general, I'm also getting ~2 seconds when comparing the `2.2.5` tag, and the `master`, `2.2.x` branches.  For me, 2 seconds is very reasonable and don't think we need to spend extra time on this.  If we can shave time down by doing easy things, great.    

That said, I'm not opposed to refactoring code to try to improve things.  I kind of want to refactor the `Query` code anyways. However, removing the above dependencies would require a fairly large reorganization of the code.  The `DataModel`, `MarvinDb`, `MarvinForm`, `Query` tool, and the `main init` codes are all quite interconnected and need `sqlalchemy` and the `ModelClasses` to build things properly.   Certainly, there are other benefits of reorganizing this code, but this will only gain us 0.3 seconds on import.  

Obviously, I don't see the problem with having dependencies.  I don't like including `from/import` statements in the middle of code wrapped in `try-excepts`.   I prefer them at the top, where I'm ok wrapping them.  We use `marvindb` in a lot of places.  That's why we initialize it in the import.  Without a real db, this object initializes as a mostly empty object.  We can probably be a bit smarter here.    

I'm against taking out the dependencies from the requirements file.  The point here is that if a user or a university downloads and sets up a database, their installed Marvin is ready to go and will automatically start using the database without setting up additional dependencies.  One solution here is to have yet another requirements file, `requirements_db.txt`, but that seems like it's getting out of hand.  (As an aside somewhat related to this, we should consider moving to [pipenv](https://docs.pipenv.org/), which is a better dependency manager, will replace req files, and has been endorsed by the Python group.)
